src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.Unix.cs

```diff
@@ -17,7 +17,15 @@ public static partial class RuntimeInformation
         public static bool IsOSPlatform(OSPlatform osPlatform)
         {
             string name = s_osPlatformName ??= Interop.Sys.GetUnixName();
-            return osPlatform.Equals(name);
+            if (osPlatform.Equals(name))
+            {
+                return true;
+            }
+            else if (name.Length == 3 && name[0] == 'O' && name[1] == 'S' && name[2] == 'X')
```

Comment:
Wow. 🥇 

```suggestion
            else if (name.Length == 3 && name[0] == 'O' && name[1] == 'S' && name[2] == 'X') // perf optimization; this is a hot path
```

Reply:
This optimization is pointless once you have paid for OrdinalIgnoreCase comparison in `osPlatform.Equals`

Reply:
Also, I think it does not look right that the comparison for everything else is case-insensitive; but the OSX match is case sensitive.

Reply:
Ah, the OSX is coming from PAL, so we know that it will be all upper case.

If we want to make this API to be high-performance so that people do not ever have a reason to cache the result to avoid the performance overhead, I think we should cache the result of the "is current platform" check in the constructor.

Reply:
I expect that the performance of the API is uneven due to case-insensitive comparison, e.g.:
- Checking for macOS on Linux is more expensive because both macOS and Linux have 5 characters
- Checking for Windows on Linux is cheaper because we will take early path out due to different string lengths

Caching the result in constructor would make the performance consistent.

Reply:
> If we want to make this API to be high-performance so that people do not ever have a reason to cache the result to avoid the performance overhead, I think we should cache the result of the "is current platform" check in the constructor.

Is something like this what you're imagining, @jkotas?

**RuntimeInformation.Unix.cs** (platform-specific, to be defined on Windows and Browser as well)
``` c#
public static partial class RuntimeInformation
{
    internal static IsOSPlatformOrEquivalent(string osPlatform)
    {
        string name = s_osPlatformName ??= Interop.Sys.GetUnixName();

        if (osPlatform.Equals(name))
        {
            return true;
        }

        if (name.Equals("OSX"))
        {
            return osPlatform.Equals("MACOS", StringComparison.OrdinalIgnoreCase);
        }

        return false;
    }
}
```

**OSPlatform.cs**
``` c#
public readonly struct OSPlatform : IEquatable<OSPlatform>
{
    internal bool IsCurrentOSPlatform { get; set; }

    private OSPlatform(string osPlatform)
    {
        if (osPlatform == null) throw new ArgumentNullException(nameof(osPlatform));
        if (osPlatform.Length == 0) throw new ArgumentException(SR.Argument_EmptyValue, nameof(osPlatform));

        _osPlatform = osPlatform;
        IsCurrentOSPlatform = RuntimeInformation.IsOSPlatformOrEquivalent(_osPlatform);
    }
}
```

**RuntimeInformation.cs** (cross-platform)
``` c#
public static partial class RuntimeInformation
{
    public static bool IsOSPlatform(OSPlatform osPlatform)
    {
        return osPlatform.IsCurrentOSPlatform;
    }
}
```

Reply:
Yes.

> platform-specific, to be defined on Windows and Browser as well

This should not be required to fix the throughput - the assembly can stay platform neutral for that. 

However, folding this tiny assembly into CoreLib and making it platform specific would be nice to reduce the startup costs.

Reply:
> `internal bool IsCurrentOSPlatform { get; set; }`

This should not need to be `set`. Having internal readonly field instead of property would be fine too.

Reply:
>  I think we should cache the result of the "is current platform" check in the constructor.

This approach also has some disadvantages:

* higher initialization cost - we need to perform this check for every known `OSPlatform` (10 as of today, most probably will grow over the time) when initializing [these](https://github.com/dotnet/runtime/blob/911e62bbe9558abd45bd67482e3634c1532219bc/src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/OSPlatform.cs#L12-L31) static fields
* won't make much difference for scenarios where consumer library is using `OSPlatform.Create` This is common for all libraries that target .NET Standard 2.0 and don't publish netcoreapp-specific packages.

To defend my proposal:

* the requirement of `OrdinalIgnoreCase` case comparison comes from #33331 and there are some good reasons why we need that. 
* For all the use cases except checking for `macOS` on `OSX`, there is no noticeable overhead (please see the numbers that I've posted for Windows in the PR description): 
     * the "hits" like checking `RuntimeInformation.IsOSPlatform(OSPlatform.Windows)` on Windows execute the perfect hot path (all letters match)
     * the "misses": the set of known inputs is limited. Most of the strings have different lengths, so the comparison returns quickly. Even if the length is the same, the first letter is different. Again, a quick return

If the new APIs were supposed to be as fast as possible, they should rather be based on `enum` rather than `string` and should definitely not allow for an input that is concatenated name and version. The usage that I see is targeting certain versions of `Windows`, `Android`, `iOS`, and `macOS` in netcoreapp-specific apps (and every version of .NET Core supports limited number of platforms which could be just part of enumeration). I am not sure if it has any usage on Linux (does the version refers to OS or kernel version?).

Reply:
> However, folding this tiny assembly into CoreLib and making it platform specific would be nice to reduce the startup costs.

I really like this idea. Having it combined with an introduction of an enum would make it almost free to check.

Reply:
These APIs did not get much use so far, and thus their performance did not really matter.

For .NET 5+, we are making these APIs a center point of larger end-to-end feature. It is why their performance matters for .NET 5+ more than before.

> higher initialization cost - we need to perform this check for every known OSPlatform (10 as of today, most probably will grow over the time)

Right, the design of this type is not particularly lean. It has non-trivial initialization cost designed in: We have to create 10 static fields, allocate 10 string literals. To cache the current platform check, we would also compare these string literals a few times. It does not make it that much more expensive.

> won't make much difference for scenarios where consumer library is using OSPlatform.Create

As long as the platform value created by OSPlatform.Create is cached, it will make the same difference. The name of the API `OSPlatform.Create` suggests that creating the platform moniker is not cheap. I do not feel bad about making `OSPlatform.Create` more expensive, assuming that people cache the result in a singleton.

> For all the use cases except checking for macOS on OSX

I can bet that the set of platforms that form equivalence or hierarchy will grow over time.

> the new APIs were supposed to be as fast as possible, they should rather be based on enum rather than string

The libraries design team has concerns about using enums for platform in public surface. The implementation can use enums internally, but we are stuck with strings in public surface. I do not think we would gain much by using enums vs. just `bool` internally. `bool` is sufficient to optimize the check that matters - whether the platform is the current platform.

Reply:
@jkotas I've added the caching and also an xml doc remark to `.Create` method. Thank you!